### PR TITLE
PorousFlow: CO2-brine material

### PR DESCRIPTION
**Pull request by cpgr**
A complex equation of state for CO2 and brine that uses component fugacity to calculate the mutual solubility of CO2 and brine. 

It is the initial work to provide a MOOSE equivalent to the TOUGH2 ECO2N module.
brineco2_xco2l_fig
brineco2_h2og_fig

Note: this does not yet include non-isothermal effects. These will be added at a later date when I get a chance.

Refs #8818

### CODE REVIEW

**Comment by cpgr** [comment]
This PR exposed a bug in BrineFluidProperties that I'll fix shortly, see #9172

File: modules/porous_flow/src/materials/PorousFlowFluidStateFlashBase.C [hunk 1]
```diff
@@ -147,18 +178,156 @@ PorousFlowFluidStateFlashBase::computeQpProperties()
...
+  for (unsigned int v = 0; v < _num_pf_vars; ++v)
+  {
+    // Derivative of density in each phase
+    _dfluid_density_dvar[_qp][_aqueous_phase_number][v] =
+        _fsp.dfluid_density_dp[_aqueous_phase_number] *
+        _dporepressure_dvar[_qp][_aqueous_phase_number][v];
+    _dfluid_density_dvar[_qp][_aqueous_phase_number][v] +=
+        _fsp.dfluid_density_dT[_aqueous_phase_number] * _dtemperature_dvar[_qp][v];
+    _dfluid_density_dvar[_qp][_aqueous_phase_number][v] +=
+        _fsp.dfluid_density_dz[_aqueous_phase_number] * dz_dvar[v];
```
Comment (rwcarlsen) [inline modules/porous_flow/src/materials/PorousFlowFluidStateFlashBase.C]: Couldn't you also do a ``for (auto phase : phases)`` loop to remove more redundancy here and below?

And for even more deduplication (which may not be worth it) you could write a locally-scoped macro to do the dirty work for each variable (density, viscosity, etc.).

**Review by rwcarlsen — COMMENTED** [review]
Overall - looks good.  Just a few questions and small refactoring suggestions.

File: modules/porous_flow/tests/fluidstate/theis_brineco2.i [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    petsc_options = '-snes_converged_reason -ksp_diagonal_scale -ksp_diagonal_scale_fix -ksp_gmres_modifiedgramschmidt -snes_linesearch_monitor'
+    petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap -snes_atol -snes_rtol -snes_max_it'
+    petsc_options_value = 'gmres      asm      lu           NONZERO                   2               1E-8       1E-10 20'
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+  end_time = 1e5
```
Comment (andrsd) [inline modules/porous_flow/tests/fluidstate/theis_brineco2.i]: This test is marked `heavy`

Comment (rwcarlsen) [inline modules/porous_flow/tests/fluidstate/theis_brineco2.i]: [facepalm]

**Review by permcody — DISMISSED** [review]
While Robert has several good suggestions, it shouldn't hold this up from going in. I don't see any errors in this PR.

File: modules/porous_flow/include/materials/PorousFlowFluidStateFlashBase.h [hunk 3]
```diff
@@ -53,14 +53,36 @@ class PorousFlowFluidStateFlashBase : public PorousFlowVariableBase
   virtual void initQpStatefulProperties() override;
   virtual void computeQpProperties() override;
 
+  /// Data structure to pass calculated thermophysical properties
+  struct FluidStateProperties
```
Comment (cpgr) [inline modules/porous_flow/include/materials/PorousFlowFluidStateFlashBase.h]: I like this idea and the resulting code simplifications. I'll have a crack at this today - thanks @rwcarlsen

File: modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C [hunk 4]
```diff
@@ -199,16 +199,16 @@ PorousFlowFluidStateWaterNCG::thermophysicalProperties() const
   Real seff = effectiveSaturation(liquid_saturation);
   Real liquid_porepressure = _gas_porepressure[_qp] + capillaryPressure(seff);
 
-  _saturation[_qp][_aqueous_phase_number] = liquid_saturation;
-  _saturation[_qp][_gas_phase_number] = gas_saturation;
-  _porepressure[_qp][_aqueous_phase_number] = liquid_porepressure;
-  _porepressure[_qp][_gas_phase_number] = _gas_porepressure[_qp];
+  fsp.saturation[_aqueous_phase_number] = liquid_saturation;
```
Comment (cpgr) [inline modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C]: Yeah, no idea why I've passed it in.

File: modules/porous_flow/tests/fluidstate/theis_brineco2.i (hunk 2 above)
Comment (cpgr) [inline modules/porous_flow/tests/fluidstate/theis_brineco2.i]: This longish test is so we can recreate a similarity solution for a 1D radial injection problem, which holds for miscibility.

This was about as small as I could make the test while still getting a decent match. We'll use these figures in our documentation, so I thought it was worthwhile adding this test.

theis_similarity_brineco2_fig

File: modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C [hunk 5]
```diff
@@ -226,35 +226,28 @@ PorousFlowFluidStateWaterNCG::thermophysicalProperties() const
...
-  _fluid_viscosity[_qp][_gas_phase_number] = gas_viscosity;
+  fsp.fluid_density[_aqueous_phase_number] = liquid_density;
+  fsp.fluid_density[_gas_phase_number] = gas_density;
+  fsp.fluid_viscosity[_aqueous_phase_number] = liquid_viscosity;
+  fsp.fluid_viscosity[_gas_phase_number] = gas_viscosity;
 
   // Derivatives wrt PorousFlow variables are required by the kernels
   // Note: these don't need to be stateful so don't calculate them in
   // initQpStatefulProperties
   if (!_is_initqp)
```
Comment (cpgr) [inline modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C]: Not sure which `if (!_is_nodal)` you mean.

Comment (rwcarlsen) [inline modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C]: Sorry - yes I meant the early return suggestion to be:

```c++
if (!_is_initqp)
  return;
```

File: modules/porous_flow/tests/fluidstate/theis_brineco2.i (hunk 2 above)
Comment (rwcarlsen) [inline modules/porous_flow/tests/fluidstate/theis_brineco2.i]: That's fine - as @andrsd pointed out - I failed to notice that the test was market heavy.  It's fine for heavy tests to be longer.

**Review by permcody — APPROVED** [review]
This looks better. Thanks!

File: modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C (hunk 5 above)
Comment (rwcarlsen) [inline modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C]: And by `if (!_is_nodal)` section, I meant the `if (!_nodal_material)` that is later in this function at the same scope as the `if (!_is_initqp)`.  Apologies for my initial, bug-ridden comment.

Comment (cpgr) [inline modules/porous_flow/src/materials/PorousFlowFluidStateWaterNCG.C]: I think it is doing what I want it to do, but I'll double check it to make sure

**Comment by cpgr** [comment]
Thanks for the reviews

**Pull request merged**